pkg/telemetry/otel.go

```diff
@@ -65,11 +79,41 @@ func InstrumentLambdaHandler(handlerFunc interface{}) interface{} {
 
 	return otellambda.InstrumentHandler(
 		handlerFunc,
+		otellambda.WithEventToCarrier(jsonEventHeadersToCarrier),
 		otellambda.WithTracerProvider(tp),
 		otellambda.WithFlusher(asFlusher),
 	)
 }
 
+// jsonEventHeadersToCarrier returns a TextMapCarrier that extracts its data from the passed JSON event. It will look
+// for the "headers" field in the event, which is populated by the AWS API Gateway. Therefore, this function will make
+// distributed tracing work only with HTTP-based lambdas that are triggered by the API Gateway.
+func jsonEventHeadersToCarrier(eventJSON []byte) propagation.TextMapCarrier {
+	ev := map[string]any{}
+	err := json.Unmarshal(eventJSON, &ev)
+	if err != nil {
+		return propagation.MapCarrier{}
+	}
+
+	// confirm the event has a headers field
+	if _, ok := ev["headers"]; !ok {
+		return propagation.MapCarrier{}
+	}
+
+	// confirm the headers field is a map
+	headers, ok := ev["headers"].(map[string]any)
+	if !ok {
+		return propagation.MapCarrier{}
+	}
+
+	headerMap := make(map[string]string, len(headers))
+	for k, v := range headers {
+		headerMap[k] = v.(string)
+	}
+
+	return propagation.MapCarrier(headerMap)
```

Comment:
nit:

We are conforming to this event schema, yeah? https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html - if so, we could probably condense this a bit more to something like:
```go
    var event struct {
        Headers map[string]string `json:"headers"`
    }

    if err := json.Unmarshal(eventJSON, &event); err != nil {
        return propagation.MapCarrier{}
    }

    return propagation.MapCarrier(event.Headers)
```

Reply:
absolutely, great point!